### DISCUSSION TITLE
IsAlwaysEnabled - How does this FeatureInfo property work?

### DISCUSSION THREAD

**Discussion by weirdyang** [OrchardCMS/OrchardCore#16969] [Q&A] [answered]
[assembly: Feature(
    Id = "OrchardCore.WorkflowPlus.TriggerService",
    Name = "Trigger service",
    Description = "Trigger service",
    Dependencies = new[] { "OrchardCore.Workflows", "OrchardCore.WorkflowPlus", "OrchardCore.Html", "OrchardCore.Workflows.Http" },
    Category = "Workflows",
    IsAlwaysEnabled = true
)]

I have a feature entry in my `Manifest.cs` with `IsAlwaysEnabled` set to true. 
I expect it to auto enable without needing to enable in the `Features` page, however, I still needed to enable it manually.

How does this IsAlwaysEnabled work? Or what is it meant to do?

**Comment by hishamco** [comment, marked as answer]
Once you enable it, you can't disable it. So, it's always enabled

**Comment by weirdyang** [comment]
for anyone who wishes to enable certain features on startup;

```
    public sealed class EnableDefaultModulesStartupTask : ModularTenantEvents
    {
        private readonly IShellHost _shellHost;

        private readonly IShellConfiguration _shellConfiguration;

        const string ModuleSectionName = "DefaultModules";
        public EnableDefaultModulesStartupTask(
            IShellHost shellhost,
            IShellConfiguration shellConfiguration)
        {
            _shellHost = shellhost;
            _shellConfiguration = shellConfiguration;
        }
        public Task EnableFeaturesAsync()
        {
            if (_shellConfiguration
                .GetSection(ModuleSectionName)
                .Exists())
            {
                List<string> features = new();
                _shellConfiguration
                    .GetSection(ModuleSectionName)
                    .Bind(features);
                if (features.Any())
                {
                    ShellScope.AddDeferredTask(async scope =>
                    {
                        if (scope != null)
                        {
                            IShellFeaturesManager manager = scope.ServiceProvider.GetRequiredService<IShellFeaturesManager>();
                            if (features.Any())
                            {
                                IEnumerable<IFeatureInfo> availableFeatures = await manager
                                .GetAvailableFeaturesAsync();

                                IEnumerable<IFeatureInfo> activatedFeatures = await manager.GetEnabledFeaturesAsync();

                                IEnumerable<IFeatureInfo> featuresToEnable = availableFeatures
                                .Where(x =>
                                features.Any(feat => string.Equals(x.Id, feat, StringComparison.InvariantCultureIgnoreCase)))
                                .ToList();

                                IEnumerable<IFeatureInfo> check = featuresToEnable
                                .Where(feat => !activatedFeatures
                                .Any(x => x.Id == feat.Id))
                                .ToList();

                                if (check.Any())
                                {
                                    await manager.UpdateFeaturesAsync(
                                        new List<IFeatureInfo>(),
                                        check,
                                        true);
                                }
                            }
                        };
                    });
                }

            }
            return Task.CompletedTask;
        }
        public override async Task ActivatingAsync()
        {
            await EnableFeaturesAsync();
        }
    }
```

**Reply by hishamco** [reply]
Also, you can enable features on setup, please check this.

https://github.com/OrchardCoreContrib/OrchardCoreContrib.Modules/blob/b3b671b3f87085918d7adb42a94c0ec1efd504a5/src/OrchardCoreContrib.Modules.Web/Startup.cs#L17

**Reply by weirdyang** [reply]
Sometimes throw an error, using deferred task is what the works consistently, especially if your features have migrations

Also that code you shared requires updating the application, I prefer doing it via a module so I can deactivate/activate it.

**Reply by hishamco** [reply]
> Sometimes throw an error, using deferred task is what the works consistently, especially if your features have migrations

File an issue with a reproduce steps

> Also that code you shared requires updating the application, I prefer doing it via a module so I can deactivate/activate it.

Which update do you mean?

**Reply by weirdyang** [reply]
I already did and that's how I came to this solution. Another member suggested deferred task.

The code you shared is within the web application, and requires updating the web application code. I prefer to do it via module, just a preference.

**Reply by hishamco** [reply]
My code runs on the startup of the web app

**Reply by weirdyang** [reply]
Yes exactly, that what I said, it requires updating the code of the web application, I prefer to use a module so all the modifications will be confined to the module, and this makes it easier to deploy to different clients, as it just means updating a package, instead of having to update all the individual web application code. And the module has the added benefit of being easily toggled on and off via the ui.
